File: Stage16/Java/Problem4949.java
package Stage16.Java;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Problem4949 {
    static boolean isBalanced(String s) {
        Stack<Character> stack = new Stack<>();

        for (char c : s.toCharArray()) {
            if (c == '(' || c == '[') {
                stack.push(c);
            } else if (c == ')') {
                if (stack.isEmpty() || stack.peek() != '(') return false;
                stack.pop();
            } else if (c == ']') {
                if (stack.isEmpty() || stack.peek() != '[') return false;
                stack.pop();
            }
        }

        return stack.isEmpty();
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        while (true) {
            String str = br.readLine();
            if (str.equals(".")) break;
            sb.append(isBalanced(str) ? "yes\n" : "no\n");
        }

        System.out.println(sb);
        br.close();
    }
}

File: Stage16/Python/Problem4949.py
import sys
input = sys.stdin.readline

def is_balanced(s: str) -> bool:
    stack = []
    
    for ch in s:
        if ch in '([':
            stack.append(ch)
        elif ch == ')':
            if not stack or stack[-1] != '(':
                return False
            stack.pop()
        elif ch == ']':
            if not stack or stack[-1] != '[':
                return False
            stack.pop()

    return not stack

def main():

    while True:
        s = input().rstrip()
        
        if s == '.':
            break
        
        print("yes" if is_balanced(s) else "no")

if __name__ == "__main__":
    main()